### PR TITLE
KAFKA-12648: invoke exception handler for MissingSourceTopicException with named topologies

### PR DESCRIPTION
**Pull request by ableegoldman**
Followup to #11600 to invoke the streams exception handler on the MissingSourceTopicException, without killing/replacing the thread

### CODE REVIEW

**Review by guozhangwang — COMMENTED** [review]
Thanks for the follow-up PR @ableegoldman . When reviewing the first PR I thought the proposed approach is that:

1) the leader would not fail the whole assignment with the error code in the assignment, but instead it would still do the assignment on the remaining sub-topologies while encoding the missing-source-topics error in the assignment. 

2) And the followers (including the leader itself), upon receiving the assignment, could then trigger the exception handler to let users be notified about the error and decide what to do, if the decision is continue, then they will continue processing the remaining sub-topologies' assigned tasks.

3) The only temporary thing would be, before the consumer's bug is fixed, we probably want to "override" user's decision to replace the thread, i.e. we would just skip it. After the consumer's fix is in, we would still honor user's decision if they decide to replace the thread (i.e. this piece of logic would be removed).

This PR though seems not doing the same thing, especially that only the leader would trigger the user's handler and then override her decision if it's replace-thread, but other members would not be notified since the error would not be encoded, and hence their handler would not trigger. Maybe I missed some context here, could you clarify a bit?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Why do you need this dummy task ID?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java]: This is not used anywhere.

File: streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java:700]: Could you also verify that the stream thread was not replaced? You could use `KafkaStreams#metadataForLocalThreads()` for that.

**Review by cadonna — COMMENTED** [review]
@ableegoldman Thanks for the PR! 

Although there is a commit that says "exclude topologies with missing source topics from assignment" I could not find where the topologies are excluded from the assignment. 

Please find my other comments inline.

@guozhangwang What consumer bug are you referring to? Is this bug related to the stale metadata about input topics?

**Comment by guozhangwang** [comment]
@cadonna I was referring to https://issues.apache.org/jira/browse/KAFKA-13543.

I chatted with @ableegoldman a bit offline and now I get it that we want to keep up with the semantics where only fatal error code would be propagated to other members, otherwise only the leader would trigger exception handler. So I think the current high-level approach makes sense, will leave some more detailed comments.

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:532]: If we think that there will be other scenarios in the future where we want to override user's decision to skip replacing threads, then I feel this is okay. Otherwise I'm more inclined to just do it in a special condition as `decision == REPLACE THREAD && throwable instanceof MissingSourceTopicException`.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:532]: Nevermind, discussed with @ableegoldman offline and we do want to keep it a general scheme.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/TopologyMetadata.java]: +1

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:544]: I'd suggest logging it as a warn instead, also we can be more explicit in the message like "user decides to replace threads, which is not necessary; the library overrides that decision and skip.. please consider changing your exception handler class to not do so.."

**Review by guozhangwang — APPROVED** [review]
On a related note, could we update the javadoc of the `StreamsUncaughtExceptionHandler` to emphasize the following:

* For optimization purposes, certain exceptions should not result in replacing threads. If user still choose to do so, Kafka Streams will override their decisions and log a warning.
* For non-fatal exceptions, only the instance that gets the event will have its handler triggered; for fatal exceptions, all the instances will have its handler triggered and the user should always choose to close application in that case.

After the above comments are addressed, I think it's good to be merged.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: Ah, well I'm not sure if this is what you meant or not but now that you mention it, we should actually be able to associate the missing source topics with a specific subtopology -- although it doesn't make sense to do so for a specific partition, so the TaskId will still have to have a "dummy" partition

If you were asking why we need a TaskId at all, it's because that's how we correlate an exception thrown to the uncaught exception handler with a specific named topology -- the StreamsException class now has an (optional) TaskId field

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:544]: The thing is that the user _has_ to return something from the handler, there is no option to "do nothing" -- from the user perspective, the `REPLACE_THREAD` option essentially means "continue running the application", they don't care if we do so by literally replacing the thread or by just swallowing the exception. So imho it doesn't make sense to log as a warning because the user _should_ choose the `REPLACE_THREAD` option in this case.

I think the confusion is around the name "REPLACE_THREAD" when we start to consider optimizations where we don't need to actually replace the thread because we don't need to kill it in the first place, unfortunately we're stuck with the name unless we want to deprecate it and choose a better one (eg rename REPLACE_THREAD --> CONTINUE_RUNNING -- does that make sense?)

**Comment by ableegoldman** [comment]
> Although there is a commit that says "exclude topologies with missing source topics from assignment" I could not find where the topologies are excluded from the assignment.

@cadonna ah, that commit was actually from the [first PR](https://github.com/apache/kafka/pull/11600), I think it's just a relic of how I did this PR on top of that one and rebased it. Sorry for the confusion.

Otherwise I believe I've addressed all your feedback, I did need to do a bit of refactoring for [this ](https://github.com/apache/kafka/pull/11686#discussion_r788430597) but it's ultimately a small technical difference -- please lmk if there's anything else, otherwise I'll merge if the build is green tomorrow

Thanks for the reviews!

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:544]: I agree that `REPLACE_THREAD` leaks quite some internals. `CONTINUE` would be a better term. Actually, it doesn't matter what we do internally to ensure continuing the processing.

**Comment by cadonna** [comment]
> On a related note, could we update the javadoc of the StreamsUncaughtExceptionHandler to emphasize the following:

@guozhangwang As far as I see the skipping of thread replacement is only used when namend topologies are used. Named topologies are not an official feature yet, so I do not see the need to update the javadocs. I would even try to avoid it to keep them simple. Of course, when named topologies become official, we should rectify the javadocs, but maybe then we also change the semantics of `REPLACE_THREAD`/`CONTINUE` as discussed above.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java:123]: My point was that I could not see where you use the task ID in this specific case when you handle the exception and a `StreamsException` can also be created without a task ID. So, I did not see the reason to introduce a dummy task ID if it is not strictly needed by the exception. I am also fine with your current solution.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/RepartitionTopicsTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/RepartitionTopicsTest.java:186]: Could you verify that in the absence of missing input topics `topologiesWithMissingInputTopics()` and  `missingSourceTopicExceptions()` return empty collections?

**Review by cadonna — APPROVED** [review]
@ableegoldman Thanks for the updates and the clarifications!

LGTM!

I would just like to see the missing verifications in the unit tests.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java]: nit:
IntelliJ says that this could be simplified to:

```suggestion
                nonFatalExceptionsToHandle.addAll(repartitionTopics.missingSourceTopicExceptions());
```

File: streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:544]: +1 on `CONTINUE`, where the library could decide what to do actually, like whether we should replace the thread or not due to optimizations. I think it's not too late to replace it via a small KIP --- this does not need to block this PR though, just that the more I think of it, the more I feel we should make it more abstract away from implementation details.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java
Comment (ableegoldman) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java:123]: We want to add a TaskId because we want to provide information about which topology is missing the source topics, and to a lesser extent which subtopology. If we create the StreamsException without a TaskId then it's considerably more difficult to figure out which topology (or which query) is experiencing the issue -- technically you could parse the error message in the MissingSourceTopicException but that's obviously very brittle and just not a good idea. Hence, we provide the `topologyName` through a "dummy" TaskId

Comment (cadonna) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/RepartitionTopics.java:123]: Now I see that that information might be useful in the user-specified exception handler. Thank you for the clarification! Maybe we should use `-1` instead of `0` for the partition number, since the partition number doe not contain any information. Just an idea, not necessarily needed for merging.

**Comment by ableegoldman** [comment]
Test failures seem to be just flaky, triggered a rerun just to be sure

**Comment by guozhangwang** [comment]
Made a final pass on the latest commits. LGTM!

**Comment by ableegoldman** [comment]
Test failures are unrelated, merging to trunk 🥳

**Pull request merged**
